### ISSUE TITLE
evaluation on val

### ISSUE THREAD

**Issue by zzwen1** [CoinCheung/BiSeNet#85]
Your online evaluation result is very high, but I predicted on the entire validation set, and then calculated iou, the result is very low, this is why?do you make sure your evaluation code is correct?

**Comment by CoinCheung** [comment]
I am not sure, what is the correct method to compute miou please ?

**Comment by zzwen1** [comment]
Have you evaluated it on the entire val set? I tried it many times and the miou is only 0.69. Our evaluation method looks the same. I don’t know what went wrong. If you have free time, you can try .
Thank you.

**Comment by zzwen1** [comment]
> I am not sure, what is the correct method to compute miou please ?

Have you evaluated it on the entire val set? I tried it many times and the miou is only 0.69. Our evaluation method looks the same. I don’t know what went wrong. If you have free time, you can try .
Thank you.

**Comment by dzyjjpy** [comment]
@CoinCheung I met the same issue. I guess there are some bug for evaluate.py.
I tried to calculate the miou between gt_mask and pred_mask:
1.demo_dir.py to inference to get image's binary mask   (Only person: 255 and background: 0, I change in cityscapes.py)
2. A seperate cal_miou.py to calculate miou for each gt and mask pair.    

`"""
Function:  calculate the miou for person segmentation benchmark
Input:     mask_gt path, mask_pred path
Output:    mIou score
Alan Jia, Apr 26th 2020
"""

import os
import cv2
from numpy import mean
import argparse

def parse_args(argv=None):
    parser = argparse.ArgumentParser(
        description='YOLACT COCO mIOU Evaluation')
    parser.add_argument('--mask_gt_path',
                        default='/media/sdc/jiapy/datasets/cityscapes/cityscapes_insta360_psd_gt/val_bad/annotations/annotations/', type=str,    #    ./benchmark_test/mask_gt/
                        help='Mask ground truth folder path.')
    parser.add_argument('--mask_pred_path', default='/home/jiapy/workspace/segmentation/BiSeNet/bad/img/', type=str,   #   ./benchmark_test/mask_pred/
                        help='Mask prediction folder path.')
    global args
    args = parser.parse_args(argv)




"""
Funcyion:  calculate iou for one image
Input:     mask_gt, mask_pred
Output:    iou
"""
def cal_Iou(mask_gt, mask_pred):
	if mask_gt.shape != mask_pred.shape:
		print("mask_gt shape should be equal to mask_pred shape")
		exit(-1)
	count_and = 0  ## pixel number for Intersection
	count_or = 0   ## pixel number for Union
	for i in range(mask_gt.shape[0]):
		for j in range(mask_gt.shape[1]):
			if (mask_gt[i,j] > 0 and mask_pred[i,j] > 0):
				count_and = count_and + 1
			if (mask_gt[i,j] > 0 or mask_pred[i,j] > 0):
				count_or = count_or + 1
	if (count_or != 0):
		iou = count_and / float(count_or)
		print(" iou is ", iou)
	return iou

""" 
Function: calculate miou for benchmark
Input:    mask_gt_path, mask_pred_path
Output:   miou
"""
def cal_mIou(path_mask_gt, path_mask_pred):
	iou_list = []
	files = os.listdir(path_mask_gt)
	for file in files:
		print("file: ", file)
		mask_gt = cv2.imread(path_mask_gt+file, cv2.IMREAD_UNCHANGED)
		if os.path.exists(path_mask_pred+file.split('.')[0]+"_binary_mask.png"):  # _ or .
			mask_pred = cv2.imread(path_mask_pred + file.split('.')[0] + "_binary_mask.png", cv2.IMREAD_UNCHANGED)  # .
		else:
			print(" no file in pred mask")
			continue


		#if mask_gt.shape
		mask_gt_8bit = []
		mask_pred_8bit = mask_pred#[:,:,0] # get one channel
		print(mask_gt.shape)
		if len(mask_gt.shape) == 2:
			mask_gt_8bit = mask_gt
		else:
			mask_gt_8bit = mask_gt[:,:,0]

		iou = cal_Iou(mask_gt_8bit, mask_pred_8bit)
		iou_list.append(iou)
	miou = mean(iou_list)
	return miou

if __name__ == '__main__':
    parse_args()
    print(args.mask_gt_path, "  ", args.mask_pred_path)
    miou = cal_mIou(args.mask_gt_path, args.mask_pred_path)
    print("miou is ", miou)`

**Comment by CoinCheung** [comment]
@dzyjjpy Your code is not clear in my browser, what are you trying to do with it ?

**Comment by dzyjjpy** [comment]
> @dzyjjpy Your code is not clear in my browser, what are you trying to do with it ?

calculate class miou for bisenet segmentation.  (person segmentation task,  pred: 255 mean person, 0 mean background)

**Comment by CoinCheung** [comment]
Please  be more specific, what is the steps to reproduce the problem ?

**Comment by dzyjjpy** [comment]
> Please be more specific, what is the steps to reproduce the problem ?

I run evaluate.py in your code base, to get miou for all class(for my model,  background and perosn )
person's miou is not correct. pls check it

**Comment by CoinCheung** [comment]
I find no errors that harms the evaluation in the current script.  Actually I use this script to test some dataset released by a competition, and the miou is very close to the result computed by the competition submit server. 

If you find errors in the code, please point out the place where the errors lies in. If you want me help you with your own code that is developed based on this repo, please provide enough details (what is the current state and what you expect to see, and better with a piece of sample code), so that I can understand and reproduce the problem.

**Comment by zzwen1** [comment]
What is your miou on the entire val set？Have you reached 73. or more?

**Comment by CoinCheung** [comment]
What do you mean, I posted the miou tested on the val set along with the pretrained model on cityscapes. You can try the pretrained model with the command in the `README.md`
